servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/ClosureTest.java

```diff
@@ -0,0 +1,267 @@
+import io.servicetalk.concurrent.api.Completable;
+import io.servicetalk.concurrent.internal.ServiceTalkTestTimeout;
+import io.servicetalk.grpc.netty.TesterProto.Tester.BlockingTestBiDiStreamRpc;
+import io.servicetalk.grpc.netty.TesterProto.Tester.BlockingTestRequestStreamRpc;
+import io.servicetalk.grpc.netty.TesterProto.Tester.BlockingTestResponseStreamRpc;
+import io.servicetalk.grpc.netty.TesterProto.Tester.BlockingTestRpc;
+import io.servicetalk.grpc.netty.TesterProto.Tester.BlockingTesterService;
+import io.servicetalk.grpc.netty.TesterProto.Tester.TestBiDiStreamRpc;
+import io.servicetalk.grpc.netty.TesterProto.Tester.TestRequestStreamRpc;
+import io.servicetalk.grpc.netty.TesterProto.Tester.TestResponseStreamRpc;
+import io.servicetalk.grpc.netty.TesterProto.Tester.TestRpc;
+import io.servicetalk.grpc.netty.TesterProto.Tester.TesterService;
+import io.servicetalk.transport.api.ServerContext;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Collection;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.servicetalk.concurrent.api.Completable.completed;
+import static io.servicetalk.concurrent.api.Completable.defer;
+import static io.servicetalk.grpc.netty.GrpcServers.forAddress;
+import static io.servicetalk.grpc.netty.TesterProto.Tester.ServiceFactory;
+import static io.servicetalk.transport.netty.internal.AddressUtils.localAddress;
+import static java.util.Arrays.asList;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+@RunWith(Parameterized.class)
+public class ClosureTest {
+    @Rule
+    public final Timeout timeout = new ServiceTalkTestTimeout();
+
+    private final boolean closeGracefully;
+
+    public ClosureTest(final boolean closeGracefully) {
+        this.closeGracefully = closeGracefully;
+    }
+
+    @Parameterized.Parameters(name = "graceful? => {0}")
+    public static Collection<Boolean> data() {
+        return asList(true, false);
+    }
+
+    @Test
+    public void serviceImplIsClosed() throws Exception {
+        CloseSignal signal = new CloseSignal(4);
+        TesterService svc = setupCloseMock(mock(TesterService.class), signal);
+        startServerAndClose(new ServiceFactory(svc), signal);
+        verifyClosure(svc, 4 /* 4 rpc methods */);
+        signal.verifyCloseAtLeastCount(closeGracefully);
+    }
+
+    @Test
+    public void blockingServiceImplIsClosed() throws Exception {
+        CloseSignal signal = new CloseSignal(1);
```

Comment:
This also should be `4` :)

Reply:
Hmm actually I take it back. This and the one in `serviceImplIsClosed()` should be `1` as we only expect 1 close call but due to current implementation detail it is called `4` times (as you correctly pointed out)